src/core/public/plugins/plugins_service.ts

```diff
@@ -50,22 +50,35 @@ export interface PluginsServiceStart {
  */
 export class PluginsService implements CoreService<PluginsServiceSetup, PluginsServiceStart> {
   /** Plugin wrappers in topological order. */
-  private readonly plugins: Map<
-    PluginName,
-    PluginWrapper<unknown, Record<string, unknown>>
-  > = new Map();
+  private readonly plugins = new Map<PluginName, PluginWrapper<unknown, Record<string, unknown>>>();
+  private readonly pluginDependencies = new Map<PluginName, PluginName[]>();
+
   private readonly satupPlugins: PluginName[] = [];
 
   constructor(private readonly coreContext: CoreContext) {}
 
-  public async setup(deps: PluginsServiceSetupDeps) {
+  public setPluginDependencies(plugins: Array<{ id: PluginName; plugin: DiscoveredPlugin }>) {
+    // Setup map of dependencies
+    const allPluginNames = new Set<PluginName>(plugins.map(p => p.id));
+    plugins.forEach(({ id, plugin }) =>
+      this.pluginDependencies.set(id, [
+        ...plugin.requiredPlugins,
+        ...plugin.optionalPlugins.filter(optPlugin => allPluginNames.has(optPlugin)),
+      ])
+    );
+
+    return this.pluginDependencies;
+  }
+
+  public async setup(deps: PluginsServiceSetupDeps): Promise<PluginsServiceSetup> {
+    this.setPluginDependencies(deps.injectedMetadata.getPlugins());
```

Comment:
hm. this method was already called, just before `setup` https://github.com/elastic/kibana/pull/41251/files#diff-ac3204e754680bbed67c8193da8cc11bR137
should we make `setPluginDependencies`  private and expose `pluginDependencies` as a part of `setup contract`?

Reply:
Yeah I don't love this, but the ContextService needs this dependency information _before_ plugins are actually set up (so that the plugins themselves an _use_ the ContextService). Technically context containers don't need this information until a _handler is called_ but the service doesn't have any control over when that could happen.

Reply:
I see, but shouldn't we delete this call?  dependencies has been already set